src/main/java/games/strategy/util/Util.java

```diff
@@ -149,4 +151,16 @@ public static String sha512(String input) {
 
     return p.negate();
   }
+
+  /**
+   * Workaround for http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8187708
+   * Essentially a better version of {@link Date#from(Instant)}.
+   */
+  public static Date toRealDate(final Instant instant) {
+    if (instant.getEpochSecond() < 0) {
```

Comment:
Just curious...  I realize it's unlikely we'll ever have to represent a date earlier than epoch, but what is the use case for constraining it here?  If I remove the guard clause, the following test case appears to pass:

```java
@Test
public void testToRealDateNegative() {
  final Instant instant = Instant.ofEpochMilli(Long.MIN_VALUE).minusMillis(1L);
  final Date date = new Date(Long.MIN_VALUE - 1L);
  assertEquals(date, Util.toRealDate(instant));
  Util.toRealDate(Instant.MIN);
}
```

I don't know how meaningful the result is, but it's probably not any less meaningful than the bogus date we get from the positive overflow case. :smile:

Reply:
@ssoloff 
This is related to the bug I reported.
Date stores its time in a long in milliseconds, but interprets this long as unsigned, therefore the overflow is intended!
Instant however treats the long as normal signed long, so if we get a negative value from an instant we know, that date would treat it as "unsigned", much further in the future instead of in the past, and the date would be completely incorrect.

Reply:
> Date stores its time in a long in milliseconds, but interprets this long as unsigned

Not sure I'm following you here.  I can pass a negative value to the `Date` constructor and always get a time earlier than epoch:

```java
final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd G 'at' HH:mm:ss.SSS z");
dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
System.out.println(dateFormat.format(new Date(-1L))); // negative
System.out.println(dateFormat.format(new Date(Long.MIN_VALUE))); // negative
System.out.println(dateFormat.format(new Date(Long.MIN_VALUE - 1L))); // positive after overflow
System.out.println(dateFormat.format(new Date(Long.MAX_VALUE))); // positive
System.out.println(dateFormat.format(new Date(Long.MAX_VALUE + 1L))); // negative after overflow
```

Prints:

```
1969-12-31 AD at 23:59:59.999 UTC
292269055-12-02 BC at 16:47:04.192 UTC
292278994-08-17 AD at 07:12:55.807 UTC
292278994-08-17 AD at 07:12:55.807 UTC
292269055-12-02 BC at 16:47:04.192 UTC
```

Reply:
Weird, could you double check with a different JRE? 
When I was testing earlier, I gor results in the future.

Reply:
When I run the above code on OpenJDK 1.8.0_141, the output matches the output from Oracle JDK 1.8.0_131.  What do you get when you run the above code on your system?

Note that I explicitly added the era indicator (`G`) to the output format.  Without that present, it may appear that the year is far in the future because it is not output with a negative sign in the case of years before 0.

Reply:
Seems legitimate, but if this is the case this would mean that the ban forever feature (from the lobby menu) has never actually worked, because it banned someone until a date in the far past, so they are instantly unbanned on the next login attempt. 
That's why I can hardly believe it. 
This feature was implemented back when @ veqryn worked on the lobby system.

Anyways: In this case go ahead and close this PR in favour of something you proposed in #2376

Reply:
I just tested the Ban Username Forever scenario on this branch.  No exception occurs, but no record gets written to the database because the following expression in `BannedUsernameController#addBannedUsername()` is `false` due to the incoming time being earlier than the current time:

```java
if (banTill == null || banTill.isAfter(now())) {
```

The value of `banTill` was `-292275007-02-09T19:43:48.164Z`.